File: spring-graphql-webmvc/src/test/java/com/example/demo/QueryTests.java
package com.example.demo;

import com.example.demo.gql.PostController;
import com.example.demo.gql.types.Post;
import com.example.demo.gql.types.PostStatus;
import com.example.demo.service.AuthorService;
import com.example.demo.service.PostNotFoundException;
import com.example.demo.service.PostService;
import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.graphql.GraphQlTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.graphql.execution.ErrorType;
import org.springframework.graphql.test.tester.GraphQlTester;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.*;

@GraphQlTest(PostController.class)
@Slf4j
public class QueryTests {

    @Autowired
    GraphQlTester graphQlTester;

    @MockBean
    PostService postService;

    @MockBean
    AuthorService authorService;

    @Test
    void allPosts() {
        when(postService.getAllPosts()).thenReturn(List.of(
                Post.builder().id(UUID.randomUUID().toString())
                        .title("test title")
                        .content("test content")
                        .build(),
                Post.builder().id(UUID.randomUUID().toString())
                        .title("test title2")
                        .content("test content2")
                        .build()
        ));

        var allPosts = """
                query posts{
                   allPosts{
                     id
                     title
                     content
                   }
                 }""";
        graphQlTester.document(allPosts)
                .execute()
                .path("allPosts[*].title")
                .entityList(String.class)
                .satisfies(titles -> assertThat(titles).contains("test title", "test title2"));

        verify(postService, times(1)).getAllPosts();
        verifyNoMoreInteractions(postService);
    }

    @Test
    void postById() {
        var id = UUID.randomUUID().toString();
        var post = Post.builder()
                .id(id)
                .title("Post")
                .content("Content of post one")
                .createdAt(LocalDateTime.now())
                .status(PostStatus.DRAFT)
                .build();
        when(this.postService.getPostById(anyString())).thenReturn(post);
        var postById = """
                query post($postId:String!){
                   postById(postId:$postId) {
                     id
                     title
                     content
                   }
                 }""";

        graphQlTester.document(postById)
                .variable("postId", id)
                .execute()
                .path("postById.title")
                .entity(String.class).satisfies(s -> assertThat(s).startsWith("Post"));

        verify(postService, times(1)).getPostById(anyString());
        verifyNoMoreInteractions(postService);
    }

    @Test
    void postById_notFoundException() {
        var id = UUID.randomUUID().toString();
        var post = Post.builder()
                .id(id)
                .title("Post")
                .content("Content of post one")
                .createdAt(LocalDateTime.now())
                .status(PostStatus.DRAFT)
                .build();
        var exception = new PostNotFoundException(id);
        doThrow(exception).when(this.postService).getPostById(anyString());
        var postById = """
                query post($postId:String!){
                   postById(postId:$postId) {
                     id
                     title
                     content
                   }
                 }""";

        graphQlTester.document(postById)
                .variable("postId", id)
                .execute()
                .errors()
                .expect(responseError -> responseError.getErrorType() == ErrorType.NOT_FOUND)
                .expect(responseError -> Objects.equals(responseError.getMessage(), exception.getMessage()));

        verify(postService, times(1)).getPostById(anyString());
        verifyNoMoreInteractions(postService);
    }

}

File: spring-graphql-rsocket/src/test/kotlin/com/example/demo/QueryTests.kt
package com.example.demo

import com.example.demo.gql.datafetchers.PostController
import com.example.demo.gql.types.Post
import com.ninjasquad.springmockk.MockkBean
import io.kotest.matchers.ints.shouldBeGreaterThan
import io.kotest.matchers.shouldBe
import io.mockk.coEvery
import io.mockk.coVerify
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.graphql.GraphQlTest
import org.springframework.graphql.ResponseError
import org.springframework.graphql.test.tester.GraphQlTester
import java.time.LocalDateTime
import java.util.*

@OptIn(ExperimentalCoroutinesApi::class)
@GraphQlTest(controllers = [PostController::class])
internal class QueryTests {
    @Autowired
    lateinit var graphQlTester: GraphQlTester

    @MockkBean
    lateinit var postService: PostService

    @MockkBean
    lateinit var authorService: AuthorService

    @Test
    fun `get all posts`() = runTest {
        coEvery { postService.allPosts() } returns
                flowOf(
                    Post(
                        id = UUID.randomUUID(),
                        title = "Post 1",
                        content = "Post 1 content",
                        status = PostStatus.DRAFT,
                        createdAt = LocalDateTime.now()
                    ),
                    Post(
                        id = UUID.randomUUID(),
                        title = "Post 2",
                        content = "Post 2 content",
                        status = PostStatus.DRAFT,
                        createdAt = LocalDateTime.now()
                    )
                )
        val query = " { allPosts { title content }}"
        graphQlTester!!.document(query)
            .execute()
            .path("data.allPosts[*].title")
            .entityList(String::class.java).contains("test title", "test title2")

        coVerify(exactly = 1) { postService.allPosts() }
    }

    @Test
    fun `get post by id`() = runTest {
        coEvery { postService.getPostById(any()) } returns
                Post(
                    id = UUID.randomUUID(),
                    title = "Post 1",
                    content = "Post 1 content",
                    status = PostStatus.DRAFT,
                    createdAt = LocalDateTime.now()
                )
        val query = "query postById(\$postId:String!){ postById(postId:\$postId) { title content }}"
        graphQlTester!!.document(query)
            .variable("postId", "test")
            .execute()
            .path("data.postById.title")
            .entity(String::class.java).isEqualTo("test title")

        coVerify(exactly = 1) { postService.getPostById(any()) }
    }

    @Test
    fun `get post by id when not found`() = runTest {
        val id = UUID.randomUUID()
        coEvery { postService.getPostById(any()) } throws PostNotFoundException(id)

        val query = "query postById(\$postId:String!){ postById(postId:\$postId) { title content }}"
        graphQlTester!!.document(query)
            .variable("postId", id.toString())
            .execute()
            .errors()
            .satisfy { it: List<ResponseError> ->
                it.size shouldBeGreaterThan 0
                it[0].message shouldBe "Post: $id was not found."
            }
        coVerify(exactly = 1) { postService.getPostById(any()) }
    }
}
